python/cudf/cudf/core/abc.py

```diff
@@ -17,6 +21,36 @@
     import pickle
 
 
+_tls = threading.local()
+
+
+def _get_tls_stream():
+    stream = getattr(_tls, "stream", None)
+    if stream is None:
+        _tls.stream = stream = cupy.cuda.Stream(non_blocking=True)
+    return stream
+
+
+def _ensure_on_host(is_cuda, buf):
+    if is_cuda:
+        with _get_tls_stream() as stream:
+            return memoryview(cupy.asarray(buf).get(stream=stream))
+    else:
+        return memoryview(buf)
+
+
+def _ensure_maybe_on_device(is_cuda, buf):
+    buf = memoryview(buf)
+    if is_cuda:
+        with _get_tls_stream() as stream:
+            #dbuf = rmm.DeviceBuffer(size=buf.nbytes, stream=stream.ptr)
+            dbuf = rmm.DeviceBuffer(size=buf.nbytes)
+            cupy.asarray(dbuf).set(numpy.asarray(buf), stream=stream)
```

Comment:
Doing a deeper dive on the deserialization case, it appears this line takes the bulk of the time in the deserialization case. The `cupy.asarray` and `numpy.asarray` calls are negligible. However calling `.set(...)` takes `~99%` of the time. Computing the transfer rate associated with `.set(...)` in isolation, this comes out to be `~9 GB/s`, which matches what [we have observed before]( https://github.com/rapidsai/dask-cuda/issues/106#issuecomment-664676616 ).